src/company_gui.cpp

```diff
@@ -2788,8 +2829,12 @@ static WindowDesc _buy_company_desc(
 /**
  * Show the query to buy another company.
  * @param company The company to buy.
+ * @param hostile_takeover Whether this is a hostile takeover.
  */
-void ShowBuyCompanyDialog(CompanyID company)
+void ShowBuyCompanyDialog(CompanyID company, bool hostile_takeover)
 {
-	AllocateWindowDescFront<BuyCompanyWindow>(&_buy_company_desc, company);
+	BuyCompanyWindow *w = (BuyCompanyWindow *)BringWindowToFrontById(WC_BUY_COMPANY, company);
+	if (w == nullptr) {
+		new BuyCompanyWindow(&_buy_company_desc, company, hostile_takeover);
+	}
```

Comment:
```suggestion
	if (BringWindowToFrontById(WC_BUY_COMPANY, company) == nullptr) {
		new BuyCompanyWindow(&_buy_company_desc, company, hostile_takeover);
	}
```
Not sure this is conceptually better, but the casting to `BuyCompanyWindow *` seems unneeded as nothing is done with it later on that requires the type to be anything other than `Window *`. At least replace the cast with a `static_cast`, instead of the C-style cast.

Reply:
This is a copy/paste of one of the many other places we do it like this. So I didn't actually want to touch that .. but it is such an old way of doing shit :) I will see what I can do :)